pilot/pkg/proxy/envoy/v1/gateway.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"strings"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig,
+	instancesOnThisNode []*model.ServiceInstance,
+	allServicesInTheMesh []*model.Service,
+	config model.IstioConfigStore, node model.Node) Listeners {
+
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
```

Comment:
Since you are getting instancesOnThisNode, shouldn't you be filtering gateways pinned to this instance? i.e. in teh gateway spec, there is a workload label thing. you can easily filter on this..